kythe/cxx/common/path_utils.h

```diff
@@ -18,11 +18,89 @@
 #define KYTHE_CXX_COMMON_PATH_UTILS_H_
 
 #include <string>
+#include <utility>
 
 #include "absl/strings/string_view.h"
+#include "absl/types/optional.h"
+#include "kythe/cxx/common/status_or.h"
 
 namespace kythe {
 
+/// \brief PathCleaner relativizes paths against a root using CleanPath.
+class PathCleaner {
+ public:
+  /// \brief Creates a PathCleaner using the given root.
+  /// \param root The root against which to relativize.
+  ///             Will be cleaned and made an absolute path.
+  static StatusOr<PathCleaner> Create(absl::string_view root);
+
+  /// \brief Transforms the cleaned, absolute version of `path` into a path
+  ///        relative to the configured root.
+  /// \return A path relative to root, if `path` is, else a cleaned `path` or an
+  ///         error if the current directory cannot be determined.
+  StatusOr<std::string> Relativize(absl::string_view path) const;
+
+ private:
+  explicit PathCleaner(std::string root) : root_(std::move(root)) {}
+
+  std::string root_;
+};
+
+/// \brief PathRealized relativizes paths against a root using RealPath.
+class PathRealizer {
+ public:
+  /// \brief Creates a PathCleaner using the given root.
+  /// \param root The root against which to relativize.
+  ///             Will be resolved and made an absolute path.
+  static StatusOr<PathRealizer> Create(absl::string_view root);
+
+  /// \brief Transforms the resolved, absolute version of `path` into a path
+  ///        relative to the configured root.
+  /// \return A path relative to root, if `path` is, else a resolved `path` or
```

Comment:
The formatting here's weird.

Reply:
How so? should subsequent lines not be indented?

Reply:
Sorry, looks like the wrong line is displayed. I meant to ask about the lonely "an" on the next line.

Reply:
Ah, clang-format wrapped that line weirdly, fixed.